src/physics/em/detail/MollerBhabhaInteractor.i.hh

```diff
@@ -0,0 +1,219 @@
+ * handled in code *before* the interactor is constructed.
+ */
+CELER_FUNCTION MollerBhabhaInteractor::MollerBhabhaInteractor(
+    const MollerBhabhaPointers& shared,
+    const ParticleTrackView&    particle,
+    const Real3&                inc_direction,
+    SecondaryAllocatorView&     allocate)
+    : shared_(shared)
+    , inc_energy_(particle.energy().value())
+    , inc_momentum_(particle.momentum().value())
+    , inc_direction_(inc_direction)
+    , allocate_(allocate)
+    , inc_particle_is_electron_(
+          (particle.particle_id() == shared_.electron_id) ? true : false)
+{
+    CELER_EXPECT(particle.particle_id() == shared_.electron_id
+                 || particle.particle_id() == shared_.positron_id);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample e-e- or e+e- scattering using Moller or Bhabha models, depending on
+ * the incident particle.
+ *
+ * See section 10.1.4 of the Geant4 physics reference manual (release 10.6).
+ */
+template<class Engine>
+CELER_FUNCTION Interaction MollerBhabhaInteractor::operator()(Engine& rng)
+{
+    // Allocate memory for the produced electron
+    Secondary* electron_secondary = this->allocate_(1);
+
+    if (electron_secondary == nullptr)
+    {
+        // Fail to allocate space for a secondary
+        return Interaction::from_failure();
+    }
+
+    // Temporary
+    (void)sizeof(rng);
+
+    // Maximum transferable energy to the free electron
+    real_type max_kinetic_energy;
+
+    if (inc_particle_is_electron_)
+    {
+        // Moller scattering (e-e-)
+        max_kinetic_energy = 0.5 * inc_energy_.value();
+    }
+    else
+    {
+        // Bhabha scattering (e+e-)
+        max_kinetic_energy = inc_energy_.value();
+    }
+
+    // Set up sampling parameters
+    real_type total_energy = inc_energy_.value() + shared_.electron_mass_c_sq;
+    real_type x_min = shared_.min_valid_energy_.value() / inc_energy_.value();
+    real_type x_max = max_kinetic_energy / inc_energy_.value();
+    real_type gamma = total_energy / shared_.electron_mass_c_sq;
+    real_type gamma_sq = gamma * gamma;
+    real_type beta_sq  = 1.0 - (1.0 / gamma_sq);
+    real_type x;
+    real_type z;
+    real_type rejection_function_g;
+    real_type random[2];
+
+    //// Moller scattering (e-e-) ////
+    if (inc_particle_is_electron_)
+    {
+        real_type two_gamma_term = (2.0 * gamma - 1.0) / gamma_sq;
+        real_type y              = 1.0 - x_max;
+        rejection_function_g     = 1.0 - two_gamma_term * x_max
+                               + x_max * x_max
+                                     * (1.0 - two_gamma_term
+                                        + (1.0 - two_gamma_term * y) / (y * y));
+
+        do
+        {
+            random[0] = generate_canonical(rng);
+            random[1] = generate_canonical(rng);
+
+            x = x_min * x_max / (x_min * (1.0 - random[0]) + x_max * random[0]);
```

Comment:
Rearranging this, it's just a uniform sample of 1/x between 1/x_min and 1/x_max. You can make the distribution easier to understand, and more efficient, and with less code duplication, by moving it outside the `if` since it's shared between M and B.
```c++
UniformRealDistribution<> sample_inverse_x(1/x_min, 1/x_max);
...
{
    x = 1/sample_inverse_x(rng);
}
```